Staging/adxl313 iio driver

danmois opened the pull request:
The implementation of the ADXL313 IIO driver and the addition of the IIO example for the ADXL312Z, ADXL313Z, and ADXL314Z development boards.

rbolboac commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+		}
+
+		return iio_format_value(buf, len, IIO_VAL_INT_PLUS_MICRO, 1, vals);
+	} else {
+		return -EINVAL;
+	}
```
you can remove the else and simply leave return -EINVAL; check the whole file.

rbolboac commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+			return ret;
+	}
+
+	// Set operation mode
```
nitpick: use the same comment format for line comments. check through the file.
suggestion: use /* */ for line comments (format used for Linux development)

danmois replied:
Thank you! I went through all the file and corrected the format.

rbolboac commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+	if (desc->adxl313_dev->dev_type != ID_ADXL314) {
+		ret = adxl313_enable_full_res(desc->adxl313_dev);
+		if (ret)
+			return ret;
```
memory leak: use goto error_config

rbolboac commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+
+	adxl313 = iio_adxl313->adxl313_dev;
+
+	adxl313_get_raw_xyz(adxl313, &x, &y, &z);
```
use return value

danmois replied:
Modified in all occurrences of the function.

rbolboac commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+	adxl313 = iio_adxl313->adxl313_dev;
+
+	for(uint32_t i = 0; i < samples*iio_adxl313->no_of_active_channels;) {
+		adxl313_get_raw_xyz(adxl313, &data_x, &data_y, &data_z);
```
use return value

rbolboac commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+		counter ++;
+	if (mask & NO_OS_BIT(2))
+		counter ++;
+	if (mask & NO_OS_BIT(3))
+		counter ++;
```
there are only 3 channels used, the mask will never have b3 set. you may remove these lines.

rbolboac reviewed (commented):
just a few remarks on my side

amiclaus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+#define ACCEL_AXIS_X (uint32_t) 0
+#define ACCEL_AXIS_Y (uint32_t) 1
+#define ACCEL_AXIS_Z (uint32_t) 2
+#define MEGA 1000000UL
```
just a thought: these kind for definitions might fit into a generic header file. similar to what it is done on linux:
https://github.com/torvalds/linux/blob/master/include/linux/units.h

amiclaus replied:
with https://github.com/analogdevicesinc/no-OS/pull/1394 you can use the generic header file for these specific units

danmois replied:
Thank you for the observation! However, I removed the units defines since they are not used in the file.

amiclaus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+static int adxl313_iio_update_channels(void* dev, uint32_t mask);
+static int32_t adxl313_trigger_handler(struct iio_device_data *dev_data);
+// /******************************************************************************/
+// /************************ Variable Declarations ******************************/
```
nit: remove `//`

amiclaus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+		default:
+			return -EINVAL;
+		}
+	} else {
```
no need to use `else` here.  just return directly. (same as you did for `adxl313_iio_write_scale`)

amiclaus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+		default:
+			return -EINVAL;
+		}
+
```
nit: remove line

danmois replied:
Thanks for pointing to this problem. I replaced the switch with an if here. In my opinion, it is more simple to understand this way.

amiclaus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+		}
+
+		if (ret) {
+			adxl313_set_op_mode(adxl313, op_mode);
```
is this function call supposed to be here?

danmois replied:
If adxl313_set_offset fails, the adxl313_iio_write_calibbias function should return the error code generated by adxl313_set_offset, and not by the adxl313_set_op_mode(which could be executed sucessfully). However, the chip was set to standby mode for performing the setting, even if the setting action had failed. This function call tries to set the device to the state previous to standby (could have been MEASURE).

amiclaus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+		vals[0] = adxl313_iio_odr_table[adxl313->odr][0];
+		vals[1] = adxl313_iio_odr_table[adxl313->odr][1];
+		return iio_format_value(buf, len, IIO_VAL_INT_PLUS_MICRO, 1, vals);
+	} else {
```
ditto: no need for else. return directly

amiclaus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+		}
+
+		return iio_format_value(buf, len, IIO_VAL_INT_PLUS_MICRO, 1, vals);
+	} else {
```
ditto: remove `else`

amiclaus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+			return -EINVAL;
+		}
+		return strlen(buf);
+	} else {
```
dito: remove else

amiclaus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+			return -EINVAL;
+		}
+		return strlen(buf);
+	} else {
```
ditto: remove else

amiclaus reviewed (commented):
lgtm overall, just some minor comments

amiclaus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+#define ACCEL_AXIS_X (uint32_t) 0
```
nit: shouldn't these defines be in `.h` ?

danmois replied:
They are used in static functions in the .c file. It seemed more appropriate to place them here, where they are used. The header contains only the definitions for the init and remove functions, that do not deal with the accelerometer's axes.

rbolboac commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+	.channels = adxl313_channels,
+	.pre_enable = (int32_t (*)())adxl313_iio_update_channels,
+	.trigger_handler = (int32_t (*)())adxl313_trigger_handler,
+	.read_dev = (int32_t (*)())adxl313_iio_read_samples,
```
I'm curious if you have tested this functionality. To me it does not look like it would work seeing how you have not defined on application level a read buffer which would store the data.
You can check if this is working by using the plot functionality in IIO Osc.

danmois replied:
The functionality has been tested. However, the variant that includes the buffer is the right one, thank you for pointing that out.

rbolboac commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+	.num_ch = NO_OS_ARRAY_SIZE(adxl313_channels),
+	.channels = adxl313_channels,
+	.pre_enable = (int32_t (*)())adxl313_iio_update_channels,
+	.trigger_handler = (int32_t (*)())adxl313_trigger_handler,
```
Seeing how there isn't a trigger example available of this one I would remove it until the example is available and the functionality is successfully being tested.

danmois replied:
You are right, this should be removed.

danmois commented:
V1

- Formatted comments
- Removed redundant code for return values

danmois commented:
V2

- removed the trigger handler

danmois commented:
V3

- Addition of read data buffer to the example

rbolboac commented on drivers/accel/adxl313/iio_adxl313.h:
```diff
@@ -0,0 +1,74 @@
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include "iio.h"
+#include "no_os_irq.h"
```
nitpick: I don't think this include is needed at the moment (needed only for hardware trigger)

rbolboac commented on projects/eval-adxl313z/src/platform/stm32/main.c:
```diff
@@ -62,6 +66,12 @@ int main()
 
+#ifdef IIO_EXAMPLE
+	ret = iio_example_main();
+	if (ret < 0)
```
nitpick: if (ret) syntax is being used on Linux, would be nice to use the same syntax

rbolboac commented on projects/eval-adxl313z/src/platform/stm32/main.c:
```diff
@@ -75,6 +85,12 @@ int main()
 		goto error;
 #endif
 
+#if (IIO_EXAMPLE + DUMMY_EXAMPLE > 1)
```
this won't work as expected seeing how BASIC_EXAMPLE is used instead of DUMMY_EXAMPLE and DUMMY_EXAMPLE is never defined.
replace DUMMY_EXAMPLE with BASIC_EXAMPLE.

danmois commented:
V4

- modified storagebits to 16 (maximum realbits for all devices is 13 bits)

danmois commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1200 @@
+static struct scan_type adxl313_iio_accel_scan_type = {
+	.sign = 's',
+	.realbits = 20,
+	.storagebits = 32,
```
Thank you for the observation, the values were correctly set now.

danmois commented:
V5

- removed unused include (no_os_irq.h)
- adopted Linux-like syntax for checking for errors and corrected project options in main.c (platform/stm32)

CiprianRegus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1146 @@
+
+	vals[1] = SCALE_FACTOR;
+
+	if (channel->type == IIO_ACCEL) {
```
Is this `if` necessary, since you only have `IIO_ACCEL` channels?

CiprianRegus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1146 @@
+			iio_parse_value(buf, IIO_VAL_FRACTIONAL, &vals[0], &vals[1]);
+			/* Search for the given data in the range available data
+			 * Enable/disable FULL_RES depending on selection */
+			if (vals[1] == adxl312_iio_scale_table[0])
+				return adxl313_enable_full_res(adxl313);
+			else
+				return adxl313_disable_full_res(adxl313);
+		case ID_ADXL313:
+			iio_parse_value(buf, IIO_VAL_FRACTIONAL, &vals[0], &vals[1]);
+			/* Enable/disable FULL_RES depending on selection */
+			if (vals[1] == adxl313_iio_scale_table[0])
+				return adxl313_enable_full_res(adxl313);
+			else
+				return adxl313_disable_full_res(adxl313);
```
Redundant elses.

CiprianRegus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1146 @@
+	}
+
+	if (channel->type == IIO_ACCEL) {
+		switch (channel->address) {
+		case ACCEL_AXIS_X:
+			val = adxl313->x_offset_raw * multiplier;
+			return iio_format_value(buf, len, IIO_VAL_INT, 1, &val);
+		case ACCEL_AXIS_Y:
+			val = adxl313->y_offset_raw * multiplier;
+			return iio_format_value(buf, len, IIO_VAL_INT, 1, &val);
+		case ACCEL_AXIS_Z:
+			val = adxl313->z_offset_raw * multiplier;
+			return iio_format_value(buf, len, IIO_VAL_INT, 1, &val);
```
You can have only 1 return after the `switch`, which will simplify the code a bit.

CiprianRegus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1146 @@
+	if (channel->type == IIO_ACCEL) {
+		vals[0] = adxl313_iio_odr_table[adxl313->odr][0];
+		vals[1] = adxl313_iio_odr_table[adxl313->odr][1];
+		return iio_format_value(buf, len, IIO_VAL_INT_PLUS_MICRO, 1, vals);
```
Shouldn't this  `len` be 2?

danmois replied:
Yest, this is the correct version, although it works this way (work with array).

CiprianRegus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1146 @@
+
+	iio_adxl313->active_channels = mask;
+
+	if (mask & NO_OS_BIT(0))
+		counter ++;
+	if (mask & NO_OS_BIT(1))
+		counter ++;
+	if (mask & NO_OS_BIT(2))
+		counter ++;
```
You can use `no_os_hweight8` instead.

danmois replied:
Thank you for the suggestion!

danmois commented:
V6

- removed redundant else for disabling full resolution
- use of no_os_hweight8() for updating channels
- corrected formatting value for read frequency
- simplified switch statement regarding reading offset values

amiclaus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1140 @@
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+#define ACCEL_AXIS_X (uint32_t) 0
+#define ACCEL_AXIS_Y (uint32_t) 1
```
by the way, why using cast?
also, these defines might fit in an enum.
Actually, you can use `adxl313_iio_accel_chans` inline

danmois replied:
Thank you for the suggestion, I have removed these defines and used the enum.

amiclaus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1140 @@
+	int i;
+
+	for (i = 0; i < size; i++) {
+		if (freq_tbl[i][0] == val1 && freq_tbl[i][1] == val2)
```
hmm, i think brackets should be used here:

`((freq_tbl[i][0] == val1) && (freq_tbl[i][1] == val2))`

amiclaus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1140 @@
+				return adxl313_enable_full_res(adxl313);
+			break;
+		case ID_ADXL314:
+			return ret;
```
drop `ret` variable completely and just return 0.

amiclaus commented on drivers/accel/adxl313/iio_adxl313.c:
```diff
@@ -0,0 +1,1140 @@
+			/* Set the ODR value */
+			return adxl313_set_range(adxl313, range_idx);
+		case ID_ADXL314:
+			return ret;
```
drop ret and return 0 directly.

amiclaus commented on projects/eval-adxl313z/src/platform/stm32/main.c:
```diff
@@ -75,6 +85,12 @@ int main()
+Please enable only one example and re-build the project.
+#endif
+
+
```
nit: extra line

amiclaus reviewed (commented):
some comments on my side

danmois commented:
V7

- removed ret variable where it was not needed
- placed parentheses in if for removing ambiguity
- used enum members instead of defines for identifying channels

danmois commented:
V8

- changed adxl313_iio_accel_chans enum member names to uppercase

danmois commented:
V9

- modified ADXL314 scale with the value provided by the datasheet
